### PR TITLE
Improvements in the packet_device module

### PR DESCRIPTION
**Pull request by t0mk**
##### SUMMARY

This PR improves the packet_device module.
- Bump of packet-python to 1.35.
- Changed module parameter `wait` to `wait_for_public_IPv`. It's not required, and it can be 4 or 6. When `wait_for_public_IPv: 4`, the module will wait until the new device gets public IPv4 address. Similarly for IPv6. When not set, the module will not wait for the automatic IP assignment.
- Improved handling of `state: active`. In the Packet Host API, a newly allocated device is in `provisioning` state for a few minutes, before it's `active`. Many API operation fail when the device is not yet `active`, so I made the module wait for the respective API state when `state: active` is passed to the module. I increased the timeout as this takes a few minutes.
- changed (fixed) argument 'lock' to 'locked'
- iPXE boot arguments.
- NB: with the default state `present`, the module returns right after the device-creating HTTP request succeeds.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
modules/cloud/packet/packet_device.py

##### ANSIBLE VERSION

```
ansible 2.4.0 (packet-device-rework fc8c52afac) last updated 2017/03/30 15:04:06 (GMT +300)
  config file =
  configured module search path = [u'/home/tomk/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by t0mk** [comment]
Hi, could any of the maintainers please look at this?
I rebased this PR to newest devel, and did some fixes for pep8.
This is quite important fixup of the initial version of the packet_device module. There's been several users asking for these improvements on the Packet Slack channel - now we have to refer them to this PR, so I think it would be nice to correct the version in Ansible upstream.

I am open to maintainers comments and suggestions, and happy to fix those straight away.

**Comment by t0mk** [comment]
ready_for_review

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py:477]: It seems the exception message contains all watched devices instead of not `active` device, should not that be: ```% [d.hostname for d in refreshed if d.state != 'active'])``` (if so `refreshed` must be initialized before `while` block) ?

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: `a['adress']` could be used instead of `(len(a['address']) > 0)`.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: `str` is superfluous here, it should be removed.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: `str` is superfluous here, it should be removed.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: In which case is `if resp.json():` useful ?

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: Should not it be `You specified too many hostnames` ?

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: I guess this is `seconds` but unit must be added.

When `state` is `active` and `wait_for_public_IPv` is set, then value of `wait_timeout` is used twice (hence max waiting time is 2*`wait_timeout`): it should be specified.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: When `state` is `active`, `wait_timeout` is applied and the call won't block more than `wait_timeout`. The documentation of `state`must be updated too.

**Review by pilou- — COMMENTED** [review]
Have you considered adding unit or integration tests for this module ?

**Comment by pilou-** [comment]
@t0mk Here is a list of minor issues unrelated to your commits. I mention them because this pull request already contains some cleaning and `DOCUMENTATION` fixes.

* in `listify_string_name_or_id`, method use `s.split(',')` instead of `[i for i in s.split(',')]`
* there is a problem with [`lock`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L426)/[`locked`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L428) parameter. DOCUMENTATION uses [`lock`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L73), [`EXAMPLES`](https://github.com/t0mk/ansible/blob/9ec0b5e702ebcbebdb8f185de0f8ca3e791704fc/lib/ansible/modules/cloud/packet/packet_device.py#L176) uses `locked`, code use both.
* `DOCUMENTATION`
  * `requirements` field, remove `"python >= 2.6"` (see [roadmap](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_4.rst#python-2-4-and-2-5-support-discontinuation))
  * add default values for `count` (`1`), `count_offset` (`1`) and `facility` (`ewr1)
  * `hostnames` description: use `I(count)` instead of `count` ([see](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#formatting-options))
  * `features` / `plan` / `operating_system`: if available, add a link to Packets documentation (using formatting function)
  * add in `notes` field that `check_mode` isn't supported
  * `RETURN`: `devices` isn't returned in case of failure (`devices` isn't always returned, `returned` must be updated)
  * you could remove default settings (`required: false`, `default: None`)

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (t0mk) [reply lib/ansible/modules/cloud/packet/packet_device.py]: I improved the comment in the function.
The json is useful when the error comes from constraints of the Packet resources (e.g., trying to switch-off a device in "provisioning" state).
I check for the json it defensively, in order to handle also HTTP responses that don't contain the error descriptions.

**Comment by t0mk** [comment]
@pilou- Thanks for your review! I agree with most things, and I fixed them.

Integration test is a good idea, I wanted to do it myself before, but found the testing docs a bit chaotic. I will try to put one together.

**Comment by t0mk** [comment]
I added an integration test in 23b6ecd. @pilou- I noticed you are member of the testing group, can you give me feedback on the test?

In order to run it, one needs to add packet_api_token and packet_test_project_id to integration_config.yml, and then run

```
$ ansible-test --local -v packet_device
```

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [reply lib/ansible/modules/cloud/packet/packet_device.py]: I checked again, `json()` can return `None`, you're right, this check is useful. You might want to add a `try/except ValueError` around `resp.json()` (requests [doc](http://docs.python-requests.org/en/master/api/#requests.Response.json)).

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [reply lib/ansible/modules/cloud/packet/packet_device.py:477]: When `wait_timeout` is `0` the exception message is `Waiting for state \"active\" timed out for devices: []"`, what do you think of initializing `refreshed` with `watched_devices` instead of an empty list ?

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: s/timout/timeout/

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: s/secods/seconds/

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: You could add the minimum required version (`>= 1.0`) here too.

**Review by pilou- — COMMENTED** [review]
Great ! I will review the test part before end of the week.

**Comment by t0mk** [comment]
@pilou- The test that I wrote needs valid API credentials and, at least at this point, can't be run from Ansible CI at shippable.

Should I move it to `test/integration/roles/test_packet_device` and then write a testing playbook `test/integration/packet.yml`, and write entry for the playbook to `test/integration/Makefile`?

**Comment by pilou-** [comment]
> The test that I wrote needs valid API credentials and, at least at this point, can't be run from Ansible CI at shippable.

Integration tests could be modified in order to only check that authentication fails (something like [that](https://github.com/t0mk/ansible/blob/438c2079c4c3912b19fb221a8a6f2000204352cf/test/integration/targets/setup_ec2/tasks/common.yml)). In addition, unit tests (using mock) could be used.

> Should I move it to `test/integration/roles/test_packet_device` and then write a testing playbook `test/integration/packet.yml`, and write entry for the playbook to `test/integration/Makefile`?

As far as I known, no. Playbooks in `test/integration/*.yml` are legacy (they existed before ansible-test) or special cases for network testing. New integration tests go into `test/integration/targets`. As the bot wrote, a `test/integration/targets/packet_device` should be added, it seems this file should contains something like `cloud/packet`.

**Comment by t0mk** [comment]
@pilou- That's a pretty bad news - test for auth fail does not really test functionality of the packet_device module. And IMO unit test for a cloud module has very little sense - it will never properly test the real and relevant interaction with the API. Indeed `test/units/modules/cloud` is quite sparse.

Maybe I could ask Packet people for credentials dedicated to Ansible integration tests. Are the integration tests for a particular cloud module run in the CI on every single commit, or just commits changing the particular module?

I can't imagine Packet host would be comfortable having their API hit on every single Ansible commit.

With the current situation, I think it's best if I just remove the test from this PR. It's pity because an integration test playbook (run on demand with valid credentials) would be a really good resource also for learning advanced usage of the module.

**Comment by pilou-** [comment]
> Maybe I could ask Packet people for credentials dedicated to Ansible integration tests. Are the integration tests for a particular cloud module run in the CI on every single commit, or just commits changing the particular module?

Ansible Packet tests would be run when a pull request updating the Ansible Packet module would be created or updated and other pull requests impacting all modules would be created or updated. Ansible Packet tests should not be executed when pull requests related to other modules are created or executed.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_device.py:28]: I don't believe this should have changed. The module was available in 2.3

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (gundalow) [inline lib/ansible/modules/cloud/packet/packet_device.py]: You can remove this line, you will get a CI failure, though that will go away once this PR is merged

**Comment by t0mk** [comment]
@gundalow , just fixed your latest comments.

I also removed the integration test. I will contribute it as separate PR, once we, hopefully, add credentials for Packet integration tests to Ansible CI. I am actually very curious on how to proceed with that & we have green light on that from Packet people - they're ready to provide dedicated account for Ansible integration test. 

You or @pilou-: please let me know if there's anything you'd need from my side.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [inline lib/ansible/modules/cloud/packet/packet_device.py]: Oh, this method seems a workaround of a bug in the Packet library. Indeed the library [already search](https://github.com/packethost/packet-python/blob/master/packet/baseapi.py#L83) for `error` key in response and use it to raise an exception but in fact this line is never reached because `raise_for_status` is [called before](https://github.com/packethost/packet-python/blob/master/packet/baseapi.py#L67). `ok` property [returns `False`](https://github.com/requests/requests/blob/master/requests/models.py#L687) only when `raise_for_status` raise an exception.

Comment (t0mk) [reply lib/ansible/modules/cloud/packet/packet_device.py]: Yes the error handling there is not so pretty. Maybe I will try to fix it and submit PR to packet-python. This workaround seemed less complicated than going through the exception flow in packet-python, fixing and waiting.

**Comment by t0mk** [comment]
@gundalow @pilou- I got input from last Ansible testing group meeting and the integration test problem for Packet resources seems over the scope of this PR.

Can we refocus on reviewing the fixes on the packet_device module introduced by this PR? I think it would be useful for users to have this in, and it's blocking other improvements we have planned.

If you agree with me that integration test is not a possibility here, what else I can do to get this merged? I am against unit tests, since I think they don't really add any value in cases like this, when the API is subject to change, volatile, and out of control of the test author.

At the moment, the CI fails because I removed the version_added items for new module parameters, according to @gundalow's review.

**Comment by gundalow** [comment]
+1 to getting this PR in *before* the 2.4 feature freeze. We can look at testing afterwards

**Review by gundalow — APPROVED** [review]

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (t0mk) [reply lib/ansible/modules/cloud/packet/packet_device.py]: I fixed it in https://github.com/packethost/packet-python/pull/26

**Comment by t0mk** [comment]
I squashed the commits.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [reply lib/ansible/modules/cloud/packet/packet_device.py]: It is fixed and released, should not `dig_error_json` be removed (or called only when `packet-python` version is older than 1.34) ?

**Comment by pilou-** [comment]
About tests: this module doesn't call directly the HTTP API (for example it's not possible to set endpoint), the module relies on [`packet-python`](https://pypi.python.org/pypi/packet-python) in order to call Packet API.
Hence:
1. this module could trust and rely on tests belonging to `packet-python`; if the webservice evolves with backward incompatible changes, then `packet-python` would need to be updated (but not necessarily this Ansible module).
2. unit tests make sense for this Ansible module

@t0mk you should comment this pull request with [`ready_for_review`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#cmd-ready_for_review) comment in order to get rid of [`needs_revision`](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#label-needs_revision) label. But there isn't any maintainer for this module (neither module maintainer nor namespace maintainer), a way could be to add this pull request to the [public core meeting agenda](https://github.com/ansible/community/issues/220). [Meetings](https://github.com/ansible/community/blob/master/meetings/core-team.yaml) are Tuesday 19:00 UTC and Thursday 15:00 UTC.

**Comment by t0mk** [comment]
I fixed the error handling (after my PR fixing the root cause in packet-python was accepted, and packet-python re-released), I also fixed it in all the other pending PRs for packet modules.

I am not going to write any unit test for cloud modules, because I think it's actually counterproductive. I wrote the reasons above, and in email. I do want to test the modules, but in a useful way - I put integration tests for packet modules to this repo: https://github.com/t0mk/packet-ansible-tests. I will use those for checking the stability of module after changes.

**Comment by t0mk** [comment]
ready_for_review

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (bcoca) [inline lib/ansible/modules/cloud/packet/packet_device.py]: 'manage' should cover all actions, no need to be this explicit in the description

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (abadger) [inline lib/ansible/modules/cloud/packet/packet_device.py]: When catching a generic exception, not knowing why it failed; you should include the stacktrace.  It's also safer to manually transform the exception into a strig usig to_native.  So this code will look like this:
``` python
import traceback

from ansible.module_utils._text import to_native
[...]

module.fail_json(msg='failed to set machine state %s, error: %s'
                           % (state, to_native(e)), exception=traceback.format_exc())

**Comment by abadger** [comment]
These changes break backwards compatibility.  The module is status='preview' but we've not decided how to decide if compatibility breaks are allowed for that status.  This was brought up at the IRC meeting today but no decision was reached.

* Everyone agreed that backwards incompatibilities need to be documented.  So that's a concrete change to this PR that can be made.
* Several proposals were made on how to decide:
  * Leave it to the module maintainers (so the authors of the packet_device module) to discuss and decide
  * Leave it to the domain experts (so the authors of modules that live inside of cloud/packet) to discuss and decide
  * Decide on something(?) project-wide with case-by-case exceptions

I'll put this question as those proposals onto the agenda for Tuesday.

**Comment by t0mk** [comment]
@abadger @bcoca Thanks for your review comments, I will fix them tomorrow morning european time.

Ad breakign backward compatibility: I am happy to document it. Would you have any example on how t document it? Link to a module doc or a guide?

**Comment by abadger** [comment]
@t0mk Here's one from the unarchive module https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/files/unarchive.py#L41

**Comment by t0mk** [comment]
@abadger, I fixed your comments, and documented the deprecated arguments.

I also added two options for iPXE boot of Packet devices, it's a small change.

**Comment by t0mk** [comment]
@abadger, I fixed your review comments. The exception handling cause the merge conflict.

I added options for iPXE boot for device in Packet host, it's a small change. Then I squashed the commits.

**Comment by jctanner** [comment]
@t0mk this needs a rebase

**Comment by t0mk** [comment]
@jctanner, I rebased.

**Comment by abadger** [comment]
@t0mk The above errors from shippable mean that you need to add version_added: 2.4 to the documentation for those options.

**Comment by t0mk** [comment]
@abadger I know - I cleared the version_added there because @gundalow wrote that I can remove it. I guess it was because he wasn't sure if this will go to 2.4.:
https://github.com/ansible/ansible/pull/23127#discussion_r131205111

Anyway, I re-added the version_added to the new args. I'd be grateful for concise and consistent instructions on improving this PR.

**Comment by abadger** [comment]
Code-wise this looks ready.  We'll get back to you on whether we want something more for backwards incompatibility no later than Tuesday next week (our next public IRC meeting)

**Comment by t0mk** [comment]
Thanks!

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (abadger) [reply lib/ansible/modules/cloud/packet/packet_device.py:477]: @t0mk @pilou- github doesn't always show me what's going on... were these comments addressed?

**Comment by abadger** [comment]
@t0mk I noticed one set of comments from @pilou- that don't seem to be addressed.  can you double check that for me?  Otherwise this is good to go.

Results of today's discussion on backwards compat: tldr; Nothing further is needed

We decided that when and how to break backwards compat is largely up to the maintainers with jsut two rules: (1) incompatibilities need to be doumented (which has been done here) and (2) they can't be done in a minor release.  We're getting ready to ship 2.4 from devel so merging to devel now is also fine.

File: lib/ansible/modules/cloud/packet/packet_device.py
Comment (pilou-) [reply lib/ansible/modules/cloud/packet/packet_device.py:477]: Yes, they were.

**Comment by pilou-** [comment]
Two notes:
- default value (`ewr1`) for `facility` parameter is not documented
- `lock` parameter has been removed in the code but the documentation states that this parameter has been deprecated. It seems deprecated means _it's available but don't use it_. Should not `lock` parameter documentation removed? `lock` parameter documentation could be removed and `lock` could be added as an alias of `locked`.

**Comment by t0mk** [comment]
I documented the `lock` as deprecated param, just like it was suggested by @abadger  at https://github.com/ansible/ansible/pull/23127#issuecomment-321713397 and analogically to the `copy` deprecated option here:
https://github.com/ansible/ansible/blob/ce3d1c6ba00e1a36c016575ea7d7d04b71923486/lib/ansible/modules/files/unarchive.py#L41

I removed the defualt value of facility, it's better that way.

**Comment by pilou-** [comment]
In Ansible `core`, deprecated means `available but will be removed` (for example: http://docs.ansible.com/ansible/latest/porting_guide_2.0.html#deprecated), I guess users could be surprised.

~~Anyway, let's move forward :)~~
~~ship it~~

Edited: @t0mk `copy` is still a usable parameter of `unarchive` (see 4ac135c1b5ed6a7b9e3151b33c1f38a85f527465). Hence `lock` parameter thing need to be fixed.

**Comment by t0mk** [comment]
@pilou- I changed the "lock" param to alias of "locked".

**Comment by pilou-** [comment]
Thanks !

shipit

@gundalow, @abadger could you merge this one ?

**Comment by abadger** [comment]
Merged to devel for 2.4.0

**Pull request merged**
